internal/pkg/runtime/engines/singularity/container_linux.go

```diff
@@ -1742,3 +1713,56 @@ func (c *container) addActionsMount(system *mount.System) error {
 
 	return system.Points.AddRemount(mount.BindsTag, containerDir, flags)
 }
+
+func (c *container) prepareNetworkSetup(system *mount.System, pid int) (func() error, error) {
+	if c.netNS {
+		if os.Geteuid() == 0 && !c.userNS {
+			// we hold a reference to container network namespace
+			// by binding /proc/self/ns/net (from RPC server) inside
+			// session directory
+			if err := c.session.AddFile("/netns", nil); err != nil {
+				return nil, err
+			}
+			nspath, _ := c.session.GetPath("/netns")
+			if err := system.Points.AddBind(mount.SharedTag, "/proc/self/ns/net", nspath, 0); err != nil {
+				return nil, fmt.Errorf("could not hold network namespace reference: %s", err)
+			}
+			networks := strings.Split(c.engine.EngineConfig.GetNetwork(), ",")
+
+			cniPath := &network.CNIPath{}
+
+			if c.engine.EngineConfig.File.CniConfPath != "" {
+				cniPath.Conf = c.engine.EngineConfig.File.CniConfPath
+			} else {
+				cniPath.Conf = defaultCNIConfPath
+			}
+			if c.engine.EngineConfig.File.CniPluginPath != "" {
+				cniPath.Plugin = c.engine.EngineConfig.File.CniPluginPath
+			} else {
+				cniPath.Plugin = defaultCNIPluginPath
+			}
+
+			setup, err := network.NewSetup(networks, strconv.Itoa(pid), nspath, cniPath)
+			if err != nil {
+				return nil, fmt.Errorf("network setup failed: %s", err)
+			}
+			netargs := c.engine.EngineConfig.GetNetworkArgs()
+			if err := setup.SetArgs(netargs); err != nil {
+				return nil, fmt.Errorf("error while setting network arguments: %s", err)
+			}
+
+			return func() error {
+				setup.SetEnvPath("/bin:/sbin:/usr/bin:/usr/sbin")
+
+				if err := setup.AddNetworks(); err != nil {
+					return fmt.Errorf("%s", err)
+				}
+				c.engine.EngineConfig.Network = setup
+				return nil
+			}, nil
+		} else if c.engine.EngineConfig.GetNetwork() != "none" {
```

Comment:
Maybe this can be simplified and put up as well?

Reply:
I will simplify in fakeroot PR, because there is other checks